File: internal/registry/registry.go
// Package registry turns a manifest into an installed app.
//
// It is the first thing in the platform that composes all three lower layers:
// internal/manifest decides what an app declares, internal/wasmhost decides what
// its module actually contains, and internal/store decides what happens in
// Postgres. The registry is where a claim meets its evidence.
//
// # Everything decidable at install is decided at install
//
// That is the whole design rule here, and it is the same one manifest.Validate
// follows one layer down. A manifest promising a function the module does not
// export installs fine today and fails on somebody's first call, with an error
// that surfaces long after the person who could fix it stopped looking. So the
// registry checks the promises against the bytes, once, at the moment the two
// are both in hand.
//
// What it deliberately does NOT decide is who may call anything. That is a
// grant, resolved per request against the connecting actor, and the registry
// holds no policy of its own (invariant 1).
package registry

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"strings"

	"github.com/bees-roadhouse/hive-sandbox/internal/manifest"
	"github.com/bees-roadhouse/hive-sandbox/internal/wasmhost"
)

// Errors an install can fail with. Values rather than strings, so a caller can
// tell "this manifest is wrong" from "this module is wrong" without matching
// on prose.
var (
	// ErrMissingExport is a manifest promising a function the module does not
	// have. The manifest and the module disagree; the module wins.
	ErrMissingExport = errors.New("registry: module does not export a declared function")

	// ErrNotReactor is a module without the entrypoint a reactor needs.
	ErrNotReactor = errors.New("registry: module is not a WASI reactor")

	// ErrNoModule is an app declaring functions with no module to satisfy them.
	ErrNoModule = errors.New("registry: manifest declares functions but no module")
)

// reactorInit is the export wazero calls to initialize a reactor. A guest that
// exports `_start` instead is a command: it runs once and exits, which is not
// what an app is.
const reactorInit = "_initialize"

// CheckExports verifies a manifest's promises against what a module contains.
//
// The direction is deliberate: every function the manifest DECLARES must exist,
// and a module exporting more than the manifest mentions is fine. Extra exports
// are how a guest keeps helpers, and refusing them would make the manifest a
// second copy of the module's symbol table that somebody has to maintain.
//
// The reverse ... a declared function that does not exist ... is the one that
// becomes a runtime failure, so it is the one refused here.
func CheckExports(s manifest.Surface, exports wasmhost.Exports) error {
	// A tool tier app has exactly one function and still needs the entrypoint,
	// because it is instantiated the same way everything else is.
	if len(s.Functions) > 0 && !exports.Has(reactorInit) {
		return fmt.Errorf("%w: %s exports no %s; build with -buildmode=c-shared",
			ErrNotReactor, s.App, reactorInit)
	}

	var missing []string
	for _, fn := range s.Functions {
		if !exports.Has(fn) {
			missing = append(missing, fn)
		}
	}
	if len(missing) == 0 {
		return nil
	}
	sort.Strings(missing)

	// Name what IS there. An AI reading this error is usually one rename away,
	// and "add_entry is missing" plus a list it can compare against is a fix
	// where a bare refusal is a guess.
	names := exports.Names()
	present := make([]string, 0, len(names))
	for _, e := range names {
		if !strings.HasPrefix(e, "_") {
			present = append(present, e)
		}
	}
	sort.Strings(present)

	return fmt.Errorf("%w: %s declares %s; the module exports [%s]",
		ErrMissingExport, s.App, strings.Join(missing, ", "), strings.Join(present, ", "))
}

// SurfaceHash is the content address of a derived surface.
//
// Through encoding/json, and that is not incidental. Tool.InputSchema is a
// map[string]any, so json is what makes the bytes stable at all: it sorts map
// keys, where gob and fmt's %v hand back Go's randomized iteration order. A
// content address that is intermittently wrong is worse than none, because it
// looks right almost always.
//
// Derive is deterministic for its own reasons ... it never iterates a map, and
// both its sorts key on values Validate has already made unique ... so the pair
// is deterministic end to end, and manifest's determinism test asserts exactly
// these bytes.
func SurfaceHash(s manifest.Surface) (string, error) {
	b, err := json.Marshal(s)
	if err != nil {
		return "", fmt.Errorf("registry: encode surface: %w", err)
	}
	sum := sha256.Sum256(b)
	return hex.EncodeToString(sum[:]), nil
}

// Prepared is a manifest that has survived everything decidable before anything
// touches Postgres: validated, derived, checked against its module, and
// content-addressed.
//
// It is a value rather than a set of side effects on purpose. An installer can
// print it, diff it against what is live, and hand it to a reviewer, which is
// the same property that made SchemaPlan worth separating from the DDL that
// executes it.
type Prepared struct {
	Manifest *manifest.Manifest
	Surface  manifest.Surface
	// SurfaceHash content-addresses what this install exposes. It changes when
	// the tool or route surface changes, which is what makes "is this promotion
	// an equivalent one" answerable without a diff.
	//
	// It is meant to be PERSISTED with the build rather than recomputed on
	// read, and the difference matters more than it looks: a recomputed hash
	// says "this is what we would derive now", where a reviewer needs "this is
	// what a human approved". If the deriver changes, every historical hash
	// silently changes meaning.
	SurfaceHash string

	// DeriveVersion says which deriver produced SurfaceHash, and travels with
	// it everywhere. Without it two differing hashes are ambiguous between "the
	// app changed" and "we changed", which is the question the hash exists to
	// answer. Persist both or neither ... the schema enforces that.
	DeriveVersion int
	// ModuleHash is the content address of the wasm, or empty for an app whose
	// collections are all generated CRUD and which therefore needs no module.
	ModuleHash string
}

// Prepare runs every check that does not need a database.
//
// exports comes from wasmhost.ModuleExports and carries the hash of the module
// it was read from, so **the module hash is not a separate parameter**. It used
// to be, with nothing tying the two together ... this package's doc says the
// registry is where a claim meets its evidence, and the evidence was a
// parameter a caller could pair with any hash it liked. The check would run,
// pass, and mean nothing. Same lesson as blob's Sealed, and cheaper to fix with
// one caller than with ten.
//
// The zero Exports is how an app with no module says so, and that case is real:
// an app whose collections are all `crud: true` needs no wasm at all, because
// generated operations run host-side. It is installable with a manifest and
// nothing else, which is the cheapest possible first rung for the builder loop
// (D24).
func Prepare(m *manifest.Manifest, exports wasmhost.Exports) (Prepared, error) {
	if err := m.Validate(); err != nil {
		return Prepared{}, err
	}

	surface := m.Derive()

	if len(surface.Functions) > 0 && exports.IsZero() {
		return Prepared{}, fmt.Errorf("%w: %s declares %d", ErrNoModule, m.Name, len(surface.Functions))
	}
	if len(surface.Functions) > 0 {
		if err := CheckExports(surface, exports); err != nil {
			return Prepared{}, err
		}
	}

	hash, err := SurfaceHash(surface)
	if err != nil {
		return Prepared{}, err
	}

	return Prepared{
		Manifest: m, Surface: surface,
		SurfaceHash: hash, DeriveVersion: manifest.DeriveVersion,
		ModuleHash: exports.ModuleHash(),
	}, nil
}

// NeedsModule reports whether this app has any guest code at all.
func (p Prepared) NeedsModule() bool { return len(p.Surface.Functions) > 0 }

// InstallSpec is everything a writer needs to register this build and install
// it, in a form that carries no database handle.
//
// Same seam as SchemaPlan, one level up: the registry decides WHAT gets
// written and internal/store is the only thing that writes it, because the
// grant predicate lives there and a second package holding a pool would be a
// second package reaching Postgres with no reason to know about grants.
//
// It is also why this package imports no pgx at all. A registry that could open
// a transaction would eventually open one.
type InstallSpec struct {
	// Slug is the app name. Unique per owner, not globally.
	Slug string
	Kind string
	// ManifestJSON is stored verbatim, so what was installed can be read back
	// rather than re-derived.
	ManifestJSON []byte

	// SurfaceHash and DeriveVersion travel together or not at all. A hash with
	// no deriver is a number nobody can interpret; the schema enforces it.
	SurfaceHash   string
	DeriveVersion int

	// ModuleHash is empty for an app that needs no wasm (D24).
	ModuleHash string

	Schema manifest.SchemaPlan
}

// InstallSpec renders what to write for one OWNER. It does not decide who may
// write it: that is a grant, resolved by the caller against the connecting
// actor.
//
// The owner is a parameter because the schema name is scoped to it. Two owners
// installing the same app get separate schemas and therefore separate
// documents, which an app-scoped name would not have given them.
func (p Prepared) InstallSpec(ownerKind, ownerID string) (InstallSpec, error) {
	plan, err := p.Manifest.SchemaPlan(ownerKind, ownerID)
	if err != nil {
		return InstallSpec{}, err
	}
	spec, err := p.installSpec()
	if err != nil {
		return InstallSpec{}, err
	}
	spec.Schema = plan
	return spec, nil
}

func (p Prepared) installSpec() (InstallSpec, error) {
	raw, err := json.Marshal(p.Manifest)
	if err != nil {
		return InstallSpec{}, fmt.Errorf("registry: encode manifest: %w", err)
	}
	return InstallSpec{
		Slug:          p.Manifest.Name,
		Kind:          string(p.Manifest.Kind),
		ManifestJSON:  raw,
		SurfaceHash:   p.SurfaceHash,
		DeriveVersion: p.DeriveVersion,
		ModuleHash:    p.ModuleHash,
	}, nil
}

File: crates/hive-registry/tests/registry.rs
//! Ported from registry_test.go, evidence_test.go and module_test.go.

use std::sync::Arc;

use hive_manifest::{Collection, Function, Kind, Manifest, Storage, ToolDef};
use hive_registry::{REACTOR_INIT, RegistryError, prepare};
use hive_wasmhost::{
    BytesSource, Capability, CapabilitySet, Config, Deps, Exports, Host, Module, hash_module,
};

const HELLO: &[u8] = include_bytes!("../../hive-wasmhost/testdata/hello.wasm");

fn journalish() -> Manifest {
    Manifest {
        kind: Some(Kind::App),
        name: "journal".into(),
        version: 1,
        storage: Storage {
            collections: vec![
                Collection {
                    name: "entries".into(),
                    ..Default::default()
                },
                Collection {
                    name: "drafts".into(),
                    crud: true,
                    ..Default::default()
                },
            ],
        },
        functions: vec![
            Function {
                name: "add_entry".into(),
                doc: String::new(),
            },
            Function {
                name: "search".into(),
                doc: String::new(),
            },
        ],
        tools: vec![ToolDef {
            name: "journal.add".into(),
            function: "add_entry".into(),
            ..Default::default()
        }],
        ..Default::default()
    }
}

/// The D24 case: every collection generated, no functions, no wasm.
fn crud_only() -> Manifest {
    Manifest {
        kind: Some(Kind::App),
        name: "bookmarks".into(),
        version: 1,
        storage: Storage {
            collections: vec![Collection {
                name: "links".into(),
                crud: true,
                ..Default::default()
            }],
        },
        ..Default::default()
    }
}

async fn host() -> Host {
    Host::new(Config::default(), Deps::default())
        .await
        .expect("host")
}

/// Produces the evidence these tests check manifests against.
///
/// It goes through the real seam: encode a module, compile it on a real host,
/// read back what the host says it exports. There is deliberately no shorter
/// path. `Exports` has a private hash precisely so a caller cannot pair one
/// module's name list with another module's address, and a test-only
/// constructor to save these few milliseconds would hand that back to every
/// caller in the process.
async fn test_exports(h: &Host, names: &[&str]) -> Exports {
    let wasm = module_exporting(names);
    let hash = hash_module(&wasm);
    h.module_exports(
        &Module {
            hash,
            app: "fixture".into(),
            version: "0.0.1".into(),
            memory_pages: 1,
            ..Default::default()
        },
        Arc::new(BytesSource::new(wasm)),
    )
    .await
    .expect("module_exports")
}

/// Hand-encodes the smallest valid wasm module that exports a memory and the
/// named functions, each a no-op.
///
/// Hand-encoded rather than built with a toolchain because the interesting
/// inputs here are export NAMES ("declares add_entry, exports add_entrie") and
/// checking a rename into the build pipeline is a minute per case. The encoder
/// is boring on purpose: one type, N functions of it, one memory, the exports,
/// N empty bodies.
fn module_exporting(names: &[&str]) -> Vec<u8> {
    const SEC_TYPE: u8 = 1;
    const SEC_FUNC: u8 = 3;
    const SEC_MEMORY: u8 = 5;
    const SEC_EXPORT: u8 = 7;
    const SEC_CODE: u8 = 10;
    const EXPORT_FUNC: u8 = 0x00;
    const EXPORT_MEMORY: u8 = 0x02;

    let mut m = vec![0x00, b'a', b's', b'm', 0x01, 0x00, 0x00, 0x00];
    // One type: () -> ().
    m.extend(section(
        SEC_TYPE,
        &[uleb(1), vec![0x60, 0x00, 0x00]].concat(),
    ));
    // N functions, all of type 0.
    let mut funcs = uleb(names.len() as u32);
    funcs.extend(std::iter::repeat_n(0x00, names.len()));
    m.extend(section(SEC_FUNC, &funcs));
    // One memory, minimum one page. check_module refuses a module without
    // one, because the ABI is copies in and out of guest memory.
    m.extend(section(SEC_MEMORY, &[uleb(1), vec![0x00, 0x01]].concat()));
    let mut exports = uleb(names.len() as u32 + 1);
    exports.extend(name("memory"));
    exports.extend([EXPORT_MEMORY, 0x00]);
    for (i, n) in names.iter().enumerate() {
        exports.extend(name(n));
        exports.push(EXPORT_FUNC);
        exports.extend(uleb(i as u32));
    }
    m.extend(section(SEC_EXPORT, &exports));
    // Bodies: no locals, immediate `end`.
    let mut code = uleb(names.len() as u32);
    for _ in names {
        let body = [0x00, 0x0b];
        code.extend(uleb(body.len() as u32));
        code.extend(body);
    }
    m.extend(section(SEC_CODE, &code));
    m
}

fn section(id: u8, body: &[u8]) -> Vec<u8> {
    [vec![id], uleb(body.len() as u32), body.to_vec()].concat()
}

fn name(s: &str) -> Vec<u8> {
    [uleb(s.len() as u32), s.as_bytes().to_vec()].concat()
}

fn uleb(mut v: u32) -> Vec<u8> {
    let mut out = Vec::new();
    loop {
        let mut b = (v & 0x7f) as u8;
        v >>= 7;
        if v != 0 {
            b |= 0x80;
        }
        out.push(b);
        if v == 0 {
            return out;
        }
    }
}

// --- the fixture itself ----------------------------------------------------------

/// A fixture that quietly produces something other than what it says would
/// make every test below pass for the wrong reason.
#[tokio::test]
async fn the_fixture_encoder_produces_what_it_claims() {
    let h = host().await;
    let exports = test_exports(&h, &[REACTOR_INIT, "add_entry", "search"]).await;
    for want in [REACTOR_INIT, "add_entry", "search"] {
        assert!(
            exports.has(want),
            "the encoded module does not export {want:?}; names = {:?}",
            exports.names()
        );
    }
    assert!(
        !exports.has("never_encoded"),
        "the fixture reports an export it was never given"
    );
    assert!(
        !exports.module_hash().is_empty(),
        "evidence with no module hash is not evidence"
    );
    h.close().await;
}

/// Exports must be bound to the bytes they were read from, or the registry's
/// whole premise (a claim meeting its evidence) is a pair of parameters a
/// caller chose independently.
#[tokio::test]
async fn exports_carry_the_hash_of_their_own_module() {
    let h = host().await;
    let one = test_exports(&h, &[REACTOR_INIT, "add_entry"]).await;
    let two = test_exports(&h, &[REACTOR_INIT, "add_entry", "search"]).await;
    assert_ne!(
        one.module_hash(),
        two.module_hash(),
        "two different modules hashed the same"
    );
    assert_eq!(
        one.module_hash(),
        hash_module(&module_exporting(&[REACTOR_INIT, "add_entry"]))
    );
    h.close().await;
}

// --- prepare ---------------------------------------------------------------------

#[tokio::test]
async fn prepare_accepts_a_matching_module() {
    let h = host().await;
    let p = prepare(
        &journalish(),
        &test_exports(&h, &[REACTOR_INIT, "add_entry", "search", "some_helper"]).await,
    )
    .expect("prepare");
    assert!(!p.surface_hash.is_empty(), "no surface hash");
    assert!(
        p.needs_module(),
        "an app with functions reported needing no module"
    );
    let spec = p.install_spec("user", "alice").expect("install_spec");
    assert!(
        spec.schema.schema.starts_with("app_journal_"),
        "schema = {:?}",
        spec.schema.schema
    );
    h.close().await;
}

/// A manifest is a claim and the module is the fact. Deciding this at install
/// is the difference between a bad manifest and a failure on somebody's first
/// call.
#[tokio::test]
async fn prepare_refuses_a_missing_export() {
    let h = host().await;
    let err = prepare(
        &journalish(),
        &test_exports(&h, &[REACTOR_INIT, "add_entry"]).await,
    )
    .expect_err("prepared");
    assert!(
        matches!(err, RegistryError::MissingExport { .. }),
        "err = {err}"
    );
    // The error has to name the missing one AND what is there, because the
    // fix is usually one rename away and a bare refusal is a guess.
    let text = err.to_string();
    assert!(
        text.contains("search"),
        "does not name the missing function: {text}"
    );
    assert!(
        text.contains("add_entry"),
        "does not list what the module exports: {text}"
    );
    h.close().await;
}

/// Extra exports are fine: that is how a guest keeps helpers.
#[tokio::test]
async fn prepare_allows_extra_exports() {
    let h = host().await;
    prepare(
        &journalish(),
        &test_exports(
            &h,
            &[REACTOR_INIT, "add_entry", "search", "helper", "another"],
        )
        .await,
    )
    .expect("extra exports were refused");
    h.close().await;
}

/// A guest that exports _start is a command, not a reactor.
#[tokio::test]
async fn prepare_refuses_a_non_reactor() {
    let h = host().await;
    let err = prepare(
        &journalish(),
        &test_exports(&h, &["_start", "add_entry", "search"]).await,
    )
    .expect_err("prepared");
    assert!(
        matches!(err, RegistryError::NotReactor { .. }),
        "err = {err}"
    );
    assert!(
        err.to_string().contains("reactor"),
        "the error should say how to fix it: {err}"
    );
    h.close().await;
}

/// D24: an app whose collections are all generated CRUD needs no wasm at all.
#[test]
fn crud_only_app_needs_no_module() {
    let p = prepare(&crud_only(), &Exports::none()).expect("prepare");
    assert!(
        !p.needs_module(),
        "a manifest-only app reported needing a module"
    );
    assert!(
        p.module_hash.is_empty(),
        "module hash = {:?}",
        p.module_hash
    );
    // And it still has a usable surface, or the case is pointless.
    assert_eq!(p.surface.tools.len(), 5, "want the five generated tools");
    for tool in &p.surface.tools {
        assert_eq!(
            tool.r#impl,
            hive_manifest::Impl::GeneratedCrud,
            "{} runs on a guest that does not exist",
            tool.name
        );
    }
}

/// Declaring functions with no module is the inverse mistake and fails loudly.
#[test]
fn prepare_refuses_functions_with_no_module() {
    let err = prepare(&journalish(), &Exports::none()).expect_err("prepared");
    assert!(matches!(err, RegistryError::NoModule { .. }), "err = {err}");
}

/// The surface hash has to move when the surface moves and not otherwise.
#[tokio::test]
async fn surface_hash_tracks_the_surface() {
    let h = host().await;
    let base = prepare(
        &journalish(),
        &test_exports(&h, &[REACTOR_INIT, "add_entry", "search"]).await,
    )
    .unwrap();
    let again = prepare(
        &journalish(),
        &test_exports(&h, &[REACTOR_INIT, "add_entry", "search"]).await,
    )
    .unwrap();
    assert_eq!(
        base.surface_hash, again.surface_hash,
        "two preparations of one manifest disagree"
    );

    // A DIFFERENT module with the same surface hashes the same: it is the
    // SURFACE address, not the build's.
    let other = prepare(
        &journalish(),
        &test_exports(&h, &[REACTOR_INIT, "add_entry", "search", "a_helper"]).await,
    )
    .unwrap();
    assert_ne!(
        other.module_hash, base.module_hash,
        "the two fixtures are the same module"
    );
    assert_eq!(
        other.surface_hash, base.surface_hash,
        "the surface hash moved when only the module changed"
    );

    // Adding a tool changes it.
    let mut m = journalish();
    m.tools.push(ToolDef {
        name: "journal.find".into(),
        function: "search".into(),
        ..Default::default()
    });
    let changed = prepare(
        &m,
        &test_exports(&h, &[REACTOR_INIT, "add_entry", "search"]).await,
    )
    .unwrap();
    assert_ne!(
        changed.surface_hash, base.surface_hash,
        "the surface hash did not move when a tool was added"
    );
    h.close().await;
}

/// An invalid manifest never reaches the module check.
#[tokio::test]
async fn prepare_validates_before_checking_the_module() {
    let h = host().await;
    let mut m = journalish();
    m.name = "Journal".into(); // uppercase: refused by validate
    let err =
        prepare(&m, &test_exports(&h, &[]).await).expect_err("an invalid manifest was prepared");
    assert!(
        matches!(err, RegistryError::Manifest(_)),
        "reported a module problem for a manifest problem: {err}"
    );
    h.close().await;
}

/// A surface hash without the deriver that produced it is a number nobody can
/// interpret.
#[tokio::test]
async fn prepared_carries_the_deriver() {
    let h = host().await;
    let p = prepare(
        &journalish(),
        &test_exports(&h, &[REACTOR_INIT, "add_entry", "search"]).await,
    )
    .unwrap();
    assert_eq!(p.derive_version, hive_manifest::DERIVE_VERSION);
    assert!(!p.surface_hash.is_empty() && p.derive_version != 0);
    h.close().await;
}

/// Two owners installing the same app get separate schemas, and therefore
/// separate documents (invariant 14, the schema-name case).
#[test]
fn install_spec_is_scoped_to_the_owner() {
    let p = prepare(&crud_only(), &Exports::none()).unwrap();
    let alice = p.install_spec("user", "alice-id").unwrap();
    let bob = p.install_spec("user", "bob-id").unwrap();
    assert_ne!(
        alice.schema.schema, bob.schema.schema,
        "two owners share a schema"
    );
    // And it is stable, or re-installing would land somewhere new every time.
    let again = p.install_spec("user", "alice-id").unwrap();
    assert_eq!(again.schema.schema, alice.schema.schema);
}

// --- against the real reference guest -------------------------------------------

fn hello_module(caps: &[Capability]) -> Module {
    Module {
        hash: hash_module(HELLO),
        app: "hello".into(),
        version: "0.1.0".into(),
        memory_pages: 256,
        capabilities: CapabilitySet::new(caps),
        ..Default::default()
    }
}

#[tokio::test]
async fn prepare_against_the_real_reference_guest() {
    let h = host().await;
    let exports = h
        .module_exports(
            &hello_module(&[Capability::Log, Capability::Storage]),
            Arc::new(BytesSource::new(HELLO)),
        )
        .await
        .expect("module_exports");
    let m = Manifest {
        kind: Some(Kind::App),
        name: "hello".into(),
        version: 1,
        functions: vec![
            Function {
                name: "hello".into(),
                doc: String::new(),
            },
            Function {
                name: "store_query".into(),
                doc: String::new(),
            },
        ],
        tools: vec![ToolDef {
            name: "hello.greet".into(),
            function: "hello".into(),
            ..Default::default()
        }],
        capabilities: vec!["log".into(), "storage".into()],
        ..Default::default()
    };
    let p = prepare(&m, &exports).expect("prepare against the real guest");
    // The hash on the Prepared came from the evidence rather than from a
    // second parameter.
    assert_eq!(p.module_hash, hash_module(HELLO));
    h.close().await;
}

/// A manifest promising something the real guest does not export.
#[tokio::test]
async fn prepare_catches_a_drifted_manifest() {
    let h = host().await;
    let exports = h
        .module_exports(
            &hello_module(&[Capability::Log, Capability::Storage]),
            Arc::new(BytesSource::new(HELLO)),
        )
        .await
        .unwrap();
    let m = Manifest {
        kind: Some(Kind::App),
        name: "hello".into(),
        version: 1,
        functions: vec![
            Function {
                name: "hello".into(),
                doc: String::new(),
            },
            Function {
                name: "renamed_last_week".into(),
                doc: String::new(),
            },
        ],
        ..Default::default()
    };
    let err = prepare(&m, &exports).expect_err("prepared");
    assert!(
        matches!(err, RegistryError::MissingExport { .. }),
        "err = {err}"
    );
    h.close().await;
}

/// module_exports runs the link checks too, so a module that could never be
/// instantiated is refused at install rather than on a first call.
#[tokio::test]
async fn module_exports_refuses_an_undeclared_capability() {
    let h = host().await;
    // The reference guest imports hive_storage; this grants only log.
    let res = h
        .module_exports(
            &hello_module(&[Capability::Log]),
            Arc::new(BytesSource::new(HELLO)),
        )
        .await;
    assert!(
        res.is_err(),
        "a module importing an ungranted capability was accepted at install"
    );
    h.close().await;
}

/// The reference guest is a reactor, which is what the whole build template is
/// arranged to produce. If this fails, the toolchain changed underneath us.
#[tokio::test]
async fn reference_guest_is_a_reactor() {
    let h = host().await;
    let exports = h
        .module_exports(
            &hello_module(&[Capability::Log, Capability::Storage]),
            Arc::new(BytesSource::new(HELLO)),
        )
        .await
        .unwrap();
    assert!(
        exports.has("_initialize"),
        "the reference guest exports no _initialize; it is not a reactor"
    );
    assert!(
        !exports.has("_start"),
        "the reference guest exports _start; a reactor build should prevent that"
    );
    h.close().await;
}
